packages/loopring_v3/contracts/amm/AmplifiedAmmController.sol

```diff
@@ -38,24 +43,41 @@ contract AmplifiedAmmController is IAmmController, Claimable
         return vTokenBalancesL2;
     }
 
-    function getVirtualBalances(
-        uint96[] memory tokenBalancesL2,
-        uint96[] memory /*vTokenBalancesL2*/
+   function authorizeVirtualBalances(
+        uint96[] memory balances,
+        uint96[] memory /*vBalancesOld*/,
+        uint96[] memory vBalancesNew,
+        bytes    memory /*data*/
         )
         external
-        view
         override
-        returns (uint96[] memory)
+        returns (bool)
     {
-        // Only allow updating the virtual balances if the AF = 1
-        require(getAmplificationFactor(msg.sender) == AMPLIFICATION_FACTOR_BASE, "INVALID_OPERATION");
+        uint authorizedTimestamp = curveChangeAuthorization[msg.sender];
+        bool authorized = (authorizedTimestamp > 0) && block.timestamp >= curveChangeAuthorization[msg.sender];
+
+        // Remove authorization
+        delete curveChangeAuthorization[msg.sender];
 
-        // Just set the virtual balances to the actual balances
-        uint96[] memory vTokenBalancesL2 = new uint96[](tokenBalancesL2.length);
-        for (uint i = 0; i < tokenBalancesL2.length; i++) {
-            vTokenBalancesL2[i] = tokenBalancesL2[i];
+        // Special case: Always allow updating the virtual balances if the AF = 1
+        if (!authorized && getAmplificationFactor(msg.sender) == AMPLIFICATION_FACTOR_BASE) {
+            for (uint i = 0; i < balances.length; i++) {
+                if(vBalancesNew[i] != balances[i]) {
```

Comment:
I don't understand the logic here. Please describe when new virtual balances are allowed.

Here is my understanding:

```solidity
if (!authorized) return false;
if (vBalancesNew.length != balances.length) return false;
for (uint i = 0; i < balances.length; i++) {
  if (vBalancesNew[i] < balances[i])  return false;
}

return true;
```

Reply:
Yes I think there was a mistake in the logic. I simplified the logic similarly to how you did it here.

Updating virtual balances is only allowed when:
- `authorizeCurveChange` was called and the current timestamp is in the supported range
- `AF == 1` and the virtual balances are set to the current balances